### PR TITLE
Adds Custom property to BackchannelAuthenticationResponse

### PR DESCRIPTION
**Pull request by gislikonrad**
### The problem
We have an authenticator app that is similar to other authenticator apps where when out-of-band authentication is invoked, a code is displayed in the invoking client that needs to be verified in the authenticator app. We need a way to send the client this verification code during CIBA authentication.

### The spec
According to section 3 in the CIBA spec _it allows for some extensibility - profiles can be created that define additional parameters to be sent to, **and received from**, the Backchannel Authentication Endpoint._ 

### The proposed solution
What I did was to add the preliminary `BackchannelAuthenticationResponse` to the `BackchannelUserLoginRequest` so that it can be mutated in the `IBackchannelAuthenticationUserNotificationService`. A new dictionary property was added to the `BackchannelAuthenticationResponse` which is similar to the one in `TokenResponse`. This dictionary is then added to the DTO as JsonExtensionData to be serialized.

### More
I considered adding the raw request to the `BackchannelUserLoginRequest` aswell, to be able to respond to extra request parameters, but I might add that to a separate PR.

---

**What issue does this PR address?**
~There is no issue reported for this, but I felt that explaining my issue would be easier in code. We are running v6.2.3 of Duende.IdentityServer.~
[An issue was created](https://github.com/DuendeSoftware/IdentityServer/issues/1405) for this feature request.

**Important: Any code or remarks in your Pull Request are under the following terms:**

If You provide us with any comments, bug reports, feedback, enhancements, or modifications proposed or suggested by You for the Software, such Feedback is provided on a non-confidential basis (notwithstanding any notice to the contrary You may include in any accompanying communication), and Licensor shall have the right to use such Feedback at its discretion, including, but not limited to the incorporation of such suggested changes into the Software. You hereby grant Licensor a perpetual, irrevocable, transferable, sublicensable, nonexclusive license under all rights necessary to incorporate and use your Feedback for any purpose, including to make and sell any products and services.

(see [our license](https://duendesoftware.com/license/identityserver.pdf), section 7)

### CODE REVIEW

File: src/IdentityServer/ResponseHandling/Default/BackchannelAuthenticationResponseGenerator.cs [hunk 1]
```diff
@@ -111,6 +111,7 @@ public virtual async Task<BackchannelAuthenticationResponse> ProcessAsync(Backch
             AuthenticationContextReferenceClasses = validationResult.ValidatedRequest.AuthenticationContextReferenceClasses,
             Tenant = validationResult.ValidatedRequest.Tenant,
             IdP = validationResult.ValidatedRequest.IdP,
+            Response = response
```
Comment (josephdecock) [inline src/IdentityServer/ResponseHandling/Default/BackchannelAuthenticationResponseGenerator.cs:114]: Why are we sending the response to the user login service? If we want to return custom values in the response from the CIBA endpoint, that should be passed above (line 101).

**Review by josephdecock — COMMENTED** [review]
I think maybe we need to add a new custom validator abstraction, similar to what is done on the token endpoint. Then users could provide an implementation that would build up the Custom dictionary.

File: src/IdentityServer/Models/BackchannelUserLoginRequest.cs [hunk 2]
```diff
@@ -59,4 +60,9 @@ public class BackchannelUserLoginRequest
     /// Gets or sets the validated resources.
     /// </summary>
     public ResourceValidationResult ValidatedResources { get; set; } = default!;
+
+    /// <summary>
+    /// The response that will be sent to the client
+    /// </summary>
+    public BackchannelAuthenticationResponse Response { get; set; } = default!;
```
[the remark this answers is not in the document]
Comment (gislikonrad) [reply src/IdentityServer/Models/BackchannelUserLoginRequest.cs:67]: Passing just the Custom dictionary would be sufficient.

File: src/IdentityServer/ResponseHandling/Default/BackchannelAuthenticationResponseGenerator.cs (hunk 1 above)
Comment (gislikonrad) [reply src/IdentityServer/ResponseHandling/Default/BackchannelAuthenticationResponseGenerator.cs:114]: If the custom dictionary would just be added on line 101, then the user login service couldn't add anything to the response, i.e a code that should be shown on a page and entered in the backchannel client.

**Comment by gislikonrad** [comment]
A custom validator abstraction would also work.

**Comment by josephdecock** [comment]
Closing in favor of new PR 1497 (this branch was rebased, and the custom validator was added).

**Pull request closed without merging**